org-code-javabuilder/lib/src/main/java/org/code/javabuilder/CodeExecutionManager.java

```diff
@@ -0,0 +1,182 @@
+package org.code.javabuilder;
+
+import static org.code.protocol.LoggerNames.MAIN_LOGGER;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PrintStream;
+import java.nio.file.Files;
+import java.util.List;
+import java.util.logging.Logger;
+import org.code.protocol.*;
+import org.code.protocol.LoggerUtils.ClearStatus;
+import org.code.protocol.LoggerUtils.SessionTime;
+
+/**
+ * Manages the execution of code in a project. When started, it performs the necessary pre-execution
+ * setup and starts code execution. When complete, or when asked to exit early, it ensures the
+ * necessary post-execution cleanup steps are performed.
+ */
+public class CodeExecutionManager {
+  private final ProjectFileLoader fileLoader;
+  private final InputHandler inputHandler;
+  private final OutputAdapter outputAdapter;
+  private final ExecutionType executionType;
+  private final List<String> compileList;
+  private final JavabuilderFileManager fileManager;
+  private final LifecycleNotifier lifecycleNotifier;
+  private final CodeBuilderRunnableFactory codeBuilderRunnableFactory;
+
+  private File tempFolder;
+  private InputRedirectionStream overrideInputStream;
+  private OutputPrintStream overrideOutputStream;
+  private InputStream systemInputStream;
+  private PrintStream systemOutputStream;
+  private boolean executionInProgress;
+
+  static class CodeBuilderRunnableFactory {
+    public CodeBuilderRunnable createCodeBuilderRunnable(
+        ProjectFileLoader fileLoader,
+        OutputAdapter outputAdapter,
+        File tempFolder,
+        ExecutionType executionType,
+        List<String> compileList) {
+      return new CodeBuilderRunnable(
+          fileLoader, outputAdapter, tempFolder, executionType, compileList);
+    }
+  }
+
+  public CodeExecutionManager(
+      ProjectFileLoader fileLoader,
+      InputHandler inputHandler,
+      OutputAdapter outputAdapter,
+      ExecutionType executionType,
+      List<String> compileList,
+      JavabuilderFileManager fileManager,
+      LifecycleNotifier lifecycleNotifier) {
+    this(
+        fileLoader,
+        inputHandler,
+        outputAdapter,
+        executionType,
+        compileList,
+        fileManager,
+        lifecycleNotifier,
+        new CodeBuilderRunnableFactory());
+  }
+
+  public CodeExecutionManager(
```

Comment:
Should this constructor be private?

Reply:
Oh good call, this should be package-private (still needs to be referenced for testing)